kinto/core/resource/schema.py

```diff
@@ -74,34 +74,38 @@ class PermissionsSchema(colander.SchemaNode):
         }
 
     """
+
     def __init__(self, *args, **kwargs):
         self.known_perms = kwargs.pop('permissions', tuple())
         super(PermissionsSchema, self).__init__(*args, **kwargs)
 
-    @staticmethod
-    def schema_type():
-        return colander.Mapping(unknown='preserve')
+        for perm in self.known_perms:
+            self[perm] = self._get_node_principals(perm)
+
+    def schema_type(self):
+        if self.known_perms:
+            return colander.Mapping(unknown='raise')
+        else:
+            return colander.Mapping(unknown='preserve')
 
     def deserialize(self, cstruct=colander.null):
-        # Start by deserializing a simple mapping.
-        permissions = super(PermissionsSchema, self).deserialize(cstruct)
-
-        # In case it is optional in parent schema.
-        if permissions in (colander.null, colander.drop):
-            return permissions
-
-        # Remove potential extra children from previous deserialization.
-        self.children = []
-        for perm in permissions.keys():
-            # If know permissions is limited, then validate inline.
-            if self.known_perms:
-                colander.OneOf(choices=self.known_perms)(self, perm)
 
-            # Add a String list child node with the name of ``perm``.
-            self.add(self._get_node_principals(perm))
+        if cstruct in (colander.null, colander.drop):
+            return super(PermissionsSchema, self).deserialize(cstruct)
 
-        # End up by deserializing a mapping whose keys are now known.
-        return super(PermissionsSchema, self).deserialize(permissions)
+        # If permissions are listed, check fields and produce fancy error messages
+        elif self.known_perms:
+            for perm in cstruct:
+                colander.OneOf(choices=self.known_perms)(self, perm)
```

Comment:
This preserve the error message used previously. e.g:
"permissions in body: \"foo\" is not one of read, write, collection:create, group:create"

Without it we just get an unrecognized fields message. We can set the error massage manually, but then this can override other validation messages.